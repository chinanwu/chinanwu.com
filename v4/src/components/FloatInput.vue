<template>
  <div class="FloatInput">
    <label
      class="FloatInput__label"
      :class="{ 'FloatInput__label--float': float }"
      :id="'floatInputLabel-' + label"
      :for="'floatInput-' + label"
      :data-float="float"
    >
      {{ label }}
    </label>
    <span class="FloatInput__required" v-if="required">Required Field</span>
    <input
      :name="name"
      class="FloatInput__input"
      :id="'floatInput-' + label"
      :aria-labelledby="'floatInputLabel-' + label"
      type="text"
      v-model="inputValue"
      @focusin="handleFocusIn"
      @focusout="handleFocusOut"
      :accesskey="accessKey"
      :required="required"
    />
  </div>
</template>

<script>
export default {
  name: "FloatInput",
  props: {
    label: {
      type: String,
      required: true
    },
    name: {
      type: String,
      required: true
    },
    required: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      float: false,
      inputValue: ""
    };
  },
  emits: ["change"],
  methods: {
    handleFocusIn() {
      this.float = true;
    },
    handleFocusOut() {
      if (this.inputValue === "") {
        this.float = false;
      } else {
        this.$emit("change", this.inputValue);
      }
    }
  },
  computed: {
    accessKey() {
      return this.label[0].toLowerCase();
    }
  }
};
</script>

<style lang="less" scoped>
@import "../styling/colours.less";

.FloatInput {
  position: relative;
  margin-top: 3rem;
}

.FloatInput__label {
  position: absolute;
  font-size: 2.4rem;
  transition: all 0.2s;
  padding-left: 1rem;

  &[data-float="true"] {
    transform: translateY(-2rem);
    font-size: 1.6rem;
    padding-left: 0;
  }
}

.FloatInput__input {
  border: none;
  border-bottom: 1px solid black;
  background-color: @grey-06;
  font-size: 2.4rem;
  width: 100%;

  &[required] {
    box-shadow: none;
  }
}

.FloatInput__required {
  position: absolute;
  transform: translateY(4rem);
}
</style>
