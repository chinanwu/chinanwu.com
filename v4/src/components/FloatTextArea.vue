<template>
  <div class="FloatTextArea">
    <label
      class="FloatTextArea__label"
      :class="{ 'FloatTextArea__label--float': float }"
      :id="'floatInputLabel-' + label"
      :for="'floatInput-' + label"
      :data-float="float"
    >
      {{ label }}
    </label>
    <textarea
      class="FloatTextArea__textarea"
      :id="'floatInput-' + label"
      :aria-labelledby="'floatInputLabel-' + label"
      v-model="inputValue"
      @focusin="handleFocusIn"
      @focusout="handleFocusOut"
    />
  </div>
</template>

<script>
export default {
  name: "FloatTextArea",
  props: {
    label: {
      type: String,
      required: true
    }
  },
  data() {
    return {
      float: false,
      inputValue: ""
    };
  },
  methods: {
    handleFocusIn() {
      this.float = true;
    },
    handleFocusOut() {
      if (this.inputValue === "") {
        this.float = false;
      }
    }
  }
};
</script>

<style lang="less" scoped>
@import "../styling/colours.less";

.FloatTextArea {
  position: relative;
  margin-top: 3rem;
}

.FloatTextArea__label {
  position: absolute;
  font-size: 2.4rem;
  transition: all 0.2s;
  padding-left: 1rem;

  &[data-float="true"] {
    transform: translateY(-2rem);
    font-size: 1.6rem;
    padding-left: 0;
  }
}

.FloatTextArea__textarea {
  font-size: 2.4rem;
  width: 100%;
}
</style>
