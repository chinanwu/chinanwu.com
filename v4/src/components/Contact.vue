<template>
  <section id="contact" aria-labelledby="contactHeader">
    <header>
      <h2 id="contactHeader" class="headerLabel">Contact</h2>
    </header>

    <div class="Contact__formContainer">
      <form name="email" class="Contact__form" @submit.prevent="handleSubmit">
        <div class="Contact__formInputs">
          <FloatInput
            class="Contact__formInput"
            name="email"
            label="Name"
            required
            @change="handleNameChange"
          />
          <FloatInput
            name="email"
            label="Subject"
            @change="handleSubjectChange"
          />
        </div>

        <FloatTextArea label="Content" />

        <div class="Contact__submit">
          <input class="Contact__submitBtn" type="submit" value="Submit" />
        </div>
      </form>
    </div>

    <div class="Contact__additional">
      <h3 class="headerLabel">Or, find me on:</h3>
      <ul class="Contact__additionalLinks --uppercase">
        <li>
          <a
            class="Contact__additionalLink"
            href="https://github.com/chinanwu"
            title="Salmon (Chin-An) Wu's github page"
          >
            <img
              class="Contact__additionalLinkIcon"
              :src="require('@/assets/images/github.png')"
              alt="Github logo"
            />
            <span class="Contact__additionalLinkLabel">Github</span>
          </a>
        </li>
        <li>
          <a
            class="Contact__additionalLink"
            href="https://linkedin.com/in/chinanwu"
            title="Salmon (Chin-An) Wu's LinkedIn page"
          >
            <img
              class="Contact__additionalLinkIcon"
              :src="require('@/assets/images/linkedin.png')"
              alt="LinkedIn logo"
            />
            <span class="Contact__additionalLinkLabel">LinkedIn</span>
          </a>
        </li>
      </ul>
    </div>
  </section>
</template>

<script>
import FloatInput from "@/components/FloatInput";
import FloatTextArea from "@/components/FloatTextArea";
export default {
  name: "Contact",
  components: { FloatTextArea, FloatInput },
  data() {
    return {
      nameValue: "",
      subjectValue: "",
      bodyValue: ""
    };
  },
  methods: {
    handleSubmit() {
      const url =
        "mailto:cawudev+website@gmail.com?" +
        "?subject=" +
        this.subjectValue +
        "&body=" +
        this.bodyValue +
        "," +
        this.nameValue;
      window.open(url);
    },
    handleNameChange(e) {
      this.nameValue = e;
    },
    handleSubjectChange(e) {
      this.subjectValue = e;
    }
  }
};
</script>

<style lang="less" scoped>
@import "../styling/colours.less";

.Contact__formContainer {
  display: flex;
  justify-content: center;
}

.Contact__form {
  width: 80%;
  display: flex;
  flex-direction: column;
}

.Contact__formInputs {
  display: flex;
  margin-bottom: 2rem;
  justify-content: space-around;
}

.Contact__content {
  font-size: 1.6rem;
}

.Contact__submit {
  display: flex;
  justify-content: end;
}

.Contact__submitBtn {
  border: none;
  background-color: @grey-01;
  color: @grey-06;
  font-size: 2.6rem;
  border-radius: 1rem;
  margin-top: 1rem;
  padding: 0.5rem 2rem;
  cursor: pointer;
  transition: box-shadow 0.3s;

  &:focus,
  &:hover {
    box-shadow: 0 0 0 0.6rem rgba(51, 51, 51, 0.3);
  }

  &:active {
    box-shadow: 0 0 0 0.3rem rgba(51, 51, 51, 0.3);
  }
}

.Contact__additional {
  margin-top: 3rem;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.Contact__additionalLinks {
  display: flex;
  justify-content: center;
  width: 80%;
}

.Contact__additionalLink {
  display: flex;
  align-items: center;
  margin: 2rem;
}

.Contact__additionalLinkIcon {
  width: 32px;
  height: 32px;
  display: flex;
}

.Contact__additionalLinkLabel {
  margin-left: 1rem;
  font-size: 2rem;
}

@media only screen and (max-width: 785px) {
  .Contact__formInputs {
    flex-direction: column;
  }

  .Contact__formInput {
    margin-bottom: 2rem;
  }
}

@media only screen and (max-width: 320px) {
  .Contact__additionalLinks {
    flex-direction: column;
  }

  .Contact__additionalLink {
    margin: 1rem;
  }
}
</style>
