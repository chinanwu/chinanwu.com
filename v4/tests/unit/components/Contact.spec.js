import FloatInput from "@/components/FloatInput";
import FloatTextArea from "@/components/FloatTextArea";
import { shallowMount } from "@vue/test-utils";
import Contact from "@/components/Contact.vue";

describe("Contact", () => {
  describe("rendering", function() {
    it("renders", () => {
      const wrapper = shallowMount(Contact);
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("behaviour", () => {
    it("sets input values", () => {
      const wrapper = shallowMount(Contact);
      expect(wrapper.element).toMatchSnapshot();
      const nameValue = "test";
      const subjectValue = "subject";
      const floatInputs = wrapper.findAllComponents(FloatInput);
      floatInputs[0].vm.$emit("change", nameValue);
      floatInputs[1].vm.$emit("change", subjectValue);
      expect(wrapper.vm.nameValue).toEqual(nameValue);
      expect(wrapper.vm.subjectValue).toEqual(subjectValue);
      const bodyValue = "body";
      wrapper.findComponent(FloatTextArea).vm.$emit("change", bodyValue);
      expect(wrapper.vm.bodyValue).toEqual(bodyValue);
    });

    xit("submits", () => {
      const wrapper = shallowMount(Contact);
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#contactSubmit").trigger("submit");
    });
  });
});
