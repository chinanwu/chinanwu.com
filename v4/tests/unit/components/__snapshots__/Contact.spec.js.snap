// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact behaviour sets input values 1`] = `
<section
  aria-labelledby="contactHeader"
  id="contact"
>
  <header>
    <h2
      class="headerLabel"
      id="contactHeader"
    >
      Contact
    </h2>
  </header>
  <div
    class="Contact__formContainer"
  >
    <form
      class="Contact__form"
      name="email"
    >
      <div
        class="Contact__formInputs"
      >
        <float-input-stub
          class="Contact__formInput"
        />
        <float-input-stub />
      </div>
      <float-text-area-stub />
      <div
        class="Contact__submit"
      >
        <input
          class="Contact__submitBtn"
          id="contactSubmit"
          type="submit"
          value="Submit"
        />
      </div>
    </form>
  </div>
  <div
    class="Contact__additional"
  >
    <h3
      class="headerLabel"
    >
      Or, find me on:
    </h3>
    <ul
      class="Contact__additionalLinks --uppercase"
    >
      <li>
        <a
          class="Contact__additionalLink"
          href="https://github.com/chinanwu"
          title="Salmon (Chin-An) Wu's github page"
        >
          <img
            alt="Github logo"
            class="Contact__additionalLinkIcon"
            src=""
          />
          <span
            class="Contact__additionalLinkLabel"
          >
            Github
          </span>
        </a>
      </li>
      <li>
        <a
          class="Contact__additionalLink"
          href="https://linkedin.com/in/chinanwu"
          title="Salmon (Chin-An) Wu's LinkedIn page"
        >
          <img
            alt="LinkedIn logo"
            class="Contact__additionalLinkIcon"
            src=""
          />
          <span
            class="Contact__additionalLinkLabel"
          >
            LinkedIn
          </span>
        </a>
      </li>
    </ul>
  </div>
</section>
`;

exports[`Contact rendering renders 1`] = `
<section
  aria-labelledby="contactHeader"
  id="contact"
>
  <header>
    <h2
      class="headerLabel"
      id="contactHeader"
    >
      Contact
    </h2>
  </header>
  <div
    class="Contact__formContainer"
  >
    <form
      class="Contact__form"
      name="email"
    >
      <div
        class="Contact__formInputs"
      >
        <float-input-stub
          class="Contact__formInput"
        />
        <float-input-stub />
      </div>
      <float-text-area-stub />
      <div
        class="Contact__submit"
      >
        <input
          class="Contact__submitBtn"
          id="contactSubmit"
          type="submit"
          value="Submit"
        />
      </div>
    </form>
  </div>
  <div
    class="Contact__additional"
  >
    <h3
      class="headerLabel"
    >
      Or, find me on:
    </h3>
    <ul
      class="Contact__additionalLinks --uppercase"
    >
      <li>
        <a
          class="Contact__additionalLink"
          href="https://github.com/chinanwu"
          title="Salmon (Chin-An) Wu's github page"
        >
          <img
            alt="Github logo"
            class="Contact__additionalLinkIcon"
            src=""
          />
          <span
            class="Contact__additionalLinkLabel"
          >
            Github
          </span>
        </a>
      </li>
      <li>
        <a
          class="Contact__additionalLink"
          href="https://linkedin.com/in/chinanwu"
          title="Salmon (Chin-An) Wu's LinkedIn page"
        >
          <img
            alt="LinkedIn logo"
            class="Contact__additionalLinkIcon"
            src=""
          />
          <span
            class="Contact__additionalLinkLabel"
          >
            LinkedIn
          </span>
        </a>
      </li>
    </ul>
  </div>
</section>
`;
